src/tracer/MOM_neutral_diffusion.F90

```diff
@@ -266,44 +279,107 @@ subroutine neutral_diffusion_calc_coeffs(G, GV, h, T, S, EOS, CS)
   real, dimension(SZI_(G),SZJ_(G),SZK_(G)+1) :: Sint ! Interface S (ppt)
   real, dimension(SZI_(G),SZJ_(G),SZK_(G)+1) :: Pint ! Interface pressure (Pa)
-  real, dimension(SZI_(G),SZJ_(G),SZK_(G)+1) :: dRdT ! Interface thermal expansion coefficient (kg/m3/degC)
-  real, dimension(SZI_(G),SZJ_(G),SZK_(G)+1) :: dRdS ! Interface haline expansion coefficient (kg/m3/ppt)
+  ! Allocatable variables for discontinuous reconstructions
+  ! Note here that the underscript 't' for top interface is equivalent to 'left' edge in the PPM literature
+  real, dimension(:,:,:,:), allocatable        :: T_i     ! Top edge reconstruction of temperature (degC)
+  real, dimension(:,:,:,:), allocatable        :: S_i     ! Top edge reconstruction of salinity (ppt)
+  real, dimension(:,:,:,:), allocatable        :: dRdT_i     ! dRho/dT (kg/m3/degC) at top edge
+  real, dimension(:,:,:,:), allocatable        :: dRdS_i     ! dRho/dS (kg/m3/ppt) at top edge
+  real, dimension(:,:),     allocatable        :: ppoly_coefficients ! PPM coefficients
+  ! Allcoatable variables for continuous reconstructions
+  real, dimension(:,:,:), allocatable        :: dRdT ! dRho/dT (kg/m3/degC) at interface
+  real, dimension(:,:,:), allocatable        :: dRdS ! dRho/dS (kg/m3/ppt) at interfaces
+
+  if (CS%continuous_reconstruction) then
+    if (.not. ALLOCATED(dRdT)) ALLOCATE(dRdT(SZI_(G),SZJ_(G),SZK_(G)+1))
+    if (.not. ALLOCATED(dRdS)) ALLOCATE(dRdS(SZI_(G),SZJ_(G),SZK_(G)+1))
+    dRdT(:,:,:) = 0.
+    dRdS(:,:,:) = 0.
+  else
+    if (.not. ALLOCATED(T_i))     ALLOCATE(T_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(S_i))     ALLOCATE(S_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(dRdT_i))  ALLOCATE(dRdT_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(dRdS_i))  ALLOCATE(dRdS_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(ppoly_coefficients))  ALLOCATE(ppoly_coefficients(SZK_(G),3))
+
+    T_i(:,:,:,:) = 0.
+    S_i(:,:,:,:) = 0.
+    dRdT_i(:,:,:,:) = 0.
+    dRdS_i(:,:,:,:) = 0.
+    ppoly_coefficients(:,:) = 0.
+  endif
 
   do j = G%jsc-1, G%jec+1
     ! Interpolate state to interface
     do i = G%isc-1, G%iec+1
-      call interface_scalar(G%ke, h(i,j,:), T(i,j,:), Tint(i,j,:), 2)
-      call interface_scalar(G%ke, h(i,j,:), S(i,j,:), Sint(i,j,:), 2)
+      if (CS%continuous_reconstruction) then
+        call interface_scalar(G%ke, h(i,j,:), T(i,j,:), Tint(i,j,:), 2)
+        call interface_scalar(G%ke, h(i,j,:), S(i,j,:), Sint(i,j,:), 2)
+      else
+        call PPM_reconstruction(G%ke, h(i,j,:), T(i,j,:), T_i(i,j,:,:), ppoly_coefficients)
+        if (CS%boundary_extrapolation) then
+          call PPM_boundary_extrapolation(G%ke, h(i,j,:), T(i,j,:), T_i(i,j,:,:), ppoly_coefficients)
+        endif
+        call PPM_reconstruction(G%ke, h(i,j,:), S(i,j,:), S_i(i,j,:,:), ppoly_coefficients)
+        if (CS%boundary_extrapolation) then
+          call PPM_boundary_extrapolation(G%ke, h(i,j,:), S(i,j,:), S_i(i,j,:,:), ppoly_coefficients)
```

Comment:
That's a good point that I hadn't thought of. My main reason for wanting to do some kind of extrapolation is that it seems somewhat unphysical that you would have a large jump in the sum of hEff in the case between two columns with identical columns versus the case where one is only very slightly lighter or denser than the other because you exclude the entirety of the top and bottom cells.